### PR TITLE
Add Localization UI Tests

### PR DESCRIPTION
**Pull request by aa-stripe**
## Summary
Create a UI test that visits all the screens in the UI Examples app, and takes a screenshot of each one. These screenshots can then be gathered from DerivedData (or saved in a custom location using `xcodebuild`, to aid with localization QA, or even general QA)
Added a new UI Tests target to the `UI Examples` scheme.

## Motivation
Automate Localization QA, and be able to scale the process as we add more languages.

## Testing
Ran the UI tests with `Cmd+u` with the `UI Examples` scheme selected
Also ran on the command line with this `xcodebuild` invocation. This is the invocation that we should use in CI too:
`xcodebuild clean test -workspace "Stripe.xcworkspace" -scheme "UI Examples" -configuration "Debug" -sdk "iphonesimulator" -destination "platform=iOS Simulator,name=iPhone 8 Plus,OS=11.2" -resultBundlePath "<result directory here>" -testLanguage "<language>" -testRegion "<region>"`

A note for adding this to CI - we should put this in a for loop that goes through all the languages/regions, and then upload all the screenshots as artifacts in the CI run. A rough idea is:

```
for locale in [("de", "DE"), ("en", "US"), ("fr", "FR"), ... ]:
    language, region = locale
    output_dir = "{}_{}_{}".format(OUTPUT_PATH, language, region)
    run_command('xcodebuild clean test -workspace "Stripe.xcworkspace" -scheme "UI Examples" -configuration "Debug" -sdk "iphonesimulator" -destination "platform=iOS Simulator,name=iPhone 8 Plus,OS=11.2" -resultBundlePath "{}" -testLanguage "{}" -testRegion "{}"'.format(output_dir, language, region))
```

### CODE REVIEW

**Comment by csabol-stripe** [comment]
*note* there are known issues running this with 11.2 simulator (see: http://www.openradar.me/39348762 https://openradar.appspot.com/26320475). Running on 11.4 fixed for me

**Comment by csabol-stripe** [comment]
UI strings not covered by this scheme/script:
STPAddCardViewController:
    - "Use shipping"
    - "Use delivery"
    - "Scan card"
    - "OK"
STPPaymentMethodsViewController
    - "Loading..."
STPShippingAddressViewController
    - "Use Billing"
    - "Invalid Shipping Address"
    - "OK"
    - "Delivery"
    - "Contact"
    - "Delivery Address"
STPPaymentCardTextField
    - "ZIP"
    - "Postal"
STPAddressFieldTableViewCell
(all of these except the last two might be actually covered when specififying a different region, I only ran en_US to test)
    - "Province"
    - "County"
    - "State / Province / Region"
    - "Postal code"
    - "Country"
    - "Email"
    - "Phone"
STPPaymentMethodTableViewCell
    - "%@ Ending In %@" (format string)
STPShippingMethodTableViewCell
    - "Free"

@aa-stripe let me know how you would like to proceed. I think our options include adding some additional interactions to the script to hit some of these, creating a new scheme that explicitly hits all of these, add test schemes to Standard/Custom Integration targets to try to hit these, something else?

**Comment by csabol-stripe** [comment]
Adding @danj-stripe as a reviewer. This patch is getting quite large as I'm also incorporating updated translations from QA (since they are based on and need another round trip of screenshots). Once I get the all clear from loc qa I can try to at least separate this out into a strings update patch and ui-testing/screenshot generating patch

File: ci_scripts/screenshot_localization_tester.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+for i in en,US zh-HANS,CN de,DE es,ES it,IT ja,JP nl,NL fr,FR fi,FI nb,NO da,DK pt-BR,BR pt-PT,PT sv,SE es-LA,AR fr-CA,CA nn,NO; do
+  #statements
+  IFS=",";
+  set -- $i;
+  xcodebuild clean test -workspace "${script_dir}/../Stripe.xcworkspace" -scheme "LocalizationTester" -configuration "Debug" -sdk "iphonesimulator" -destination "platform=iOS Simulator,name=iPhone 6s,OS=11.4" -resultBundlePath "$HOME/Desktop/loc_qa/$1_$2" -testLanguage $1 -testRegion $2
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply ci_scripts/screenshot_localization_tester.sh]: makes sense!

File: LocalizationTester/MockCustomerContext.m [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        if (completion) {
+            completion(_mockCustomer, nil);
+        }
+    }
+}
+- (void)attachSourceToCustomer:(id<STPSourceProtocol>)source completion:(STPErrorBlock)completion {
+    if ([source isKindOfClass:[STPToken class]] && ((STPToken *)source).card != nil) {
+        [_mockCustomer.mockSources addObject:((STPToken *)source).card];
+    }
+    completion(nil);
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply LocalizationTester/MockCustomerContext.m]: yup, thanks

File: LocalizationTesterUITests/LocalizationTesterUITests.m [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+- (void)tearDown {
+    // Put teardown code here. This method is called after the invocation of each test method in the class.
+    [super tearDown];
+}
+
+- (void)testVisitAll {
+    XCUIApplication *app = [[XCUIApplication alloc] init];
+    XCUIElementQuery *tablesQuery = app.tables;
+
+#pragma mark -
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply LocalizationTesterUITests/LocalizationTesterUITests.m]: blegh yeah, I'll add text

File: LocalizationTesterUITests/LocalizationTesterUITests.m [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            lastCell.exists && !CGRectIsEmpty(lastCell.frame) && !CGRectContainsRect(app.windows.firstMatch.frame, lastCell.frame)) {
+            [app swipeUp];
+        } else {
+            break;
+        }
+    } while (lastCell.exists && viewPortScreen < 4); // viewPortScreen < 4 as sanity check to avoid infinite loop
+}
+
+- (void)_waitForElementToAppear:(XCUIElement *)element {
+    XCTAssert([element waitForExistenceWithTimeout:5], "An exepected element did not appear on screen: \(element)");
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply LocalizationTesterUITests/LocalizationTesterUITests.m]: copy-pasta :P

File: LocalizationTests/LocalizationTests.m [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    
+    // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
+}
+
+- (void)tearDown {
+    // Put teardown code here. This method is called after the invocation of each test method in the class.
+    [super tearDown];
+}
+
+- (void)testExample {
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply LocalizationTests/LocalizationTests.m]: should be deleted. Thanks

File: Stripe/STPAddCardViewController.m [hunk 6]
```diff
@@ -48,6 +48,8 @@ @interface STPAddCardViewController ()<
     UITableViewDelegate,
     UITableViewDataSource>
 
+@property (nonatomic) BOOL alwaysShowScanCardButton;
+@property (nonatomic) BOOL alwaysEnableDoneButton;
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Stripe/STPAddCardViewController.m]: I'm a bit confused about why this is needed as well, but `_alwaysShowScanCardButton` is not a valid ivar without it. Also matches the pattern of the other properties declared in the private header

**Comment by aa-stripe** [comment]
@csabol-stripe @danj-stripe @yuki-stripe thank you so much for taking this to the finish line!

**Pull request merged**